torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -588,6 +588,167 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce_scatter(
   throw std::runtime_error("ProcessGroupMPI does not support reduce_scatter");
 }
 
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::alltoall_base(
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
+    std::vector<int>& outputSplitSizes,
+    std::vector<int>& inputSplitSizes,
+    const AllToAllOptions& opts) {
+  checkSingleTensor(outputTensors);
+  checkSingleTensor(inputTensors);
+
+  if(outputSplitSizes.size() == 0 && inputSplitSizes.size() == 0) {
+    // We can use alltoall
+    if ((outputTensors[0].numel() != inputTensors[0].numel()) ||
+        (outputTensors[0].type() != inputTensors[0].type())) {
+      throw std::runtime_error("Tensors are not equal in size or data type");
+    }
+    checkSingleTensorHelper(inputTensors[0]);
+    checkSingleTensorHelper(outputTensors[0]);
+    if(outputTensors[0].size(0) % size_ != 0) {
+      throw std::runtime_error("Tensor's dim 0 does not divide equally across group size");
+    }
+    
+    std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this](std::unique_ptr<WorkEntry>& entry) {
+        auto srcdata = (entry->src)[0];
+        auto dstdata = (entry->dst)[0];
+        c10::DeviceGuard guard(srcdata.device());
+        std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+        MPI_CHECK(MPI_Alltoall(
```

Comment:
I'm guessing that you're assuming that both input and output tensors are non-strided? I.e., they are stored contiguously in row-major order on both sending and receiving ends? 

Typically, you don't have to make this assumption. You can create a single `MPI_Type_vector` per rank than encodes stride information and let MPI handle "flattening" of data internally. For now, just a check to make sure the data is indeed in contiguous row-major order is sufficient just so you don't get garbled outputs silently.

Reply:
@drdarshan the issue with letting MPI handle flattening of data internally is performance. We have a single MPI background thread and most MPI implementations don't optimize MPI_Type_vector and other derived data types. But like you already mentioned, the extra check might be helpful to avoid silent errors and will require checking the entire input and output tensor.

Reply:
Agree with @srinivas212 on performance aspect. The necessary check is already there in `checkSingleTensor` call. I am following the same strategy as used in other API implementation code.